proposals/handler_test.go

```diff
@@ -1402,50 +1403,191 @@ func gproposal(signer *signing.EdSigner, atxid types.ATXID,
+			ActiveSetHash: sets[n/2].Hash(),
+			Beacon:        types.Beacon{1},
+		})
+		th.mconsumer.EXPECT().IsKnown(p[n].Layer, p[n].ID()).AnyTimes()
+	}
 	pid := p2p.Peer("any")
 
 	th.mclock.EXPECT().CurrentLayer().Return(lid).AnyTimes()
 	th.mm.EXPECT().ProcessedLayer().Return(lid - 2).AnyTimes()
-	th.mclock.EXPECT().LayerToTime(gomock.Any())
+	th.mclock.EXPECT().LayerToTime(gomock.Any()).AnyTimes()
 	th.mf.EXPECT().RegisterPeerHashes(pid, gomock.Any()).AnyTimes()
-	th.mf.EXPECT().GetActiveSet(gomock.Any(), set.Hash()).DoAndReturn(
-		func(_ context.Context, got types.Hash32) error {
-			require.NoError(t, activesets.Add(th.db, got, &types.EpochActiveSet{
-				Epoch: lid.GetEpoch(),
-				Set:   set,
-			}))
-			for _, id := range set {
-				th.atxsdata.AddAtx(lid.GetEpoch(), id, &atxsdata.ATX{Node: types.NodeID{1}})
-			}
-			return nil
-		},
-	)
+	type asReq struct {
+		id  types.Hash32
+		err error
+	}
+	startChs := make(map[types.Hash32]chan struct{})
+	asCh := make(chan asReq, 1) // buffered, no wait
+	for _, set := range sets {
+		th.mf.EXPECT().GetActiveSet(gomock.Any(), set.Hash()).DoAndReturn(
+			func(ctx context.Context, got types.Hash32) error {
+				startCh, found := startChs[got]
+				if found {
+					select {
+					case <-ctx.Done():
+						return ctx.Err()
+					case startCh <- struct{}{}:
+					}
+				}
+				var req asReq
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case req = <-asCh:
+				}
+				require.Equal(t, got, req.id)
+				if req.err != nil {
+					return req.err
+				}
+				require.NoError(t, activesets.Add(th.db, got, &types.EpochActiveSet{
+					Epoch: lid.GetEpoch(),
+					Set:   set,
+				}))
+				for _, id := range set {
+					th.atxsdata.AddAtx(lid.GetEpoch(), id, &atxsdata.ATX{Node: types.NodeID{1}})
+				}
+				return nil
+			},
+		).AnyTimes()
+	}
 	th.mf.EXPECT().GetAtxs(gomock.Any(), gomock.Any()).AnyTimes()
 	th.mf.EXPECT().GetBallots(gomock.Any(), gomock.Any()).AnyTimes()
 	th.mockSet.decodeAnyBallots()
 	th.mv.EXPECT().CheckEligibility(gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes().Return(nil)
 	th.mm.EXPECT().AddBallot(gomock.Any(), gomock.Any()).AnyTimes()
 	th.mm.EXPECT().AddTXsFromProposal(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes()
 
-	th.mconsumer.EXPECT().IsKnown(good.Layer, good.ID())
-	th.mconsumer.EXPECT().OnProposal(gomock.Eq(&good))
-	err = th.HandleSyncedProposal(context.Background(), good.ID().AsHash32(), pid, codec.MustEncode(&good))
-	require.NoError(t, err)
+	t.Run("non-concurrent fetching of ActiveSets", func(t *testing.T) {
+		asCh <- asReq{id: sets[0].Hash()}
+		th.mconsumer.EXPECT().OnProposal(gomock.Eq(p[0]))
+		err = th.HandleSyncedProposal(context.Background(), p[0].ID().AsHash32(), pid, codec.MustEncode(p[0]))
+		require.NoError(t, err)
+
+		th.mconsumer.EXPECT().OnProposal(gomock.Eq(p[1]))
+		err = th.HandleSyncedProposal(context.Background(), p[1].ID().AsHash32(), pid, codec.MustEncode(p[1]))
+		require.NoError(t, err)
```

Comment:
How about instead of pre-generating lists of sets and proposals and using p[0], p[1], and so on in respective tests, create only the required set and proposals in the test body?

Also, the `EXPECT().GetActiveSet` could be much simpler and explicit (expect to be called an exact number of times) for this test. Currently, it's convoluted and hard to tell whether `fetch.GetActiveSet()` is expected to be called once or multiple times (it's also not checked by the test).

Consider something like:
```suggestion
		set := types.ATXIDList{{1}, {2}}
		var p []types.Proposal
		for _, atx := range set {
			proposals = append(proposals, gproposal(t, signer, set[0], lid, &types.EpochData{
				ActiveSetHash: set.Hash(),
				Beacon:        types.Beacon{1},
			}))

		th.mconsumer.EXPECT().OnProposal(gomock.Eq(p[0]))
		// NOTE: the tests expects it to be called only once (no `AnyTimes()` anymore)
		th.mf.EXPECT().GetActiveSet(gomock.Any(), set.Hash()).DoAndReturn(
			func(ctx context.Context, got types.Hash32) error {
				require.NoError(t, activesets.Add(th.db, got, &types.EpochActiveSet{
					Epoch: lid.GetEpoch(),
					Set:   set,
				}))
				for _, id := range set {
					th.atxsdata.AddAtx(lid.GetEpoch(), id, &atxsdata.ATX{Node: types.NodeID{1}})
				}
				return nil
		})
		
		err = th.HandleSyncedProposal(context.Background(), p[0].ID().AsHash32(), pid, codec.MustEncode(p[0]))
		require.NoError(t, err)

		th.mconsumer.EXPECT().OnProposal(gomock.Eq(p[1]))
		err = th.HandleSyncedProposal(context.Background(), p[1].ID().AsHash32(), pid, codec.MustEncode(p[1]))
		require.NoError(t, err)
		// maybe also something like:
		require.True(t, th.mockController.Satisfied())
```

Reply:
The point of `GetActiveSet` mock handler is to be able to pause fetching so that multiple request processing for the same activeset can be checked. For the simpler one-request-after-another case, a simpler impl can be used, but the more complicated one will still have to be kept, so that it can be used for the concurrent cases. And the proposals are already pre-generated, just earlier in the outer test

Reply:
Without `asCh <- asReq{id: sets[0].Hash()}`, `GetActiveSet` doesn't go ahead, but I'll try to ensure it's called only once

Reply:
There are no concurrent calls to `HandleSyncedProposal` in this test case. I'm proposing not to try a generic approach that would fit all test cases, but create simple and explicit expectations per test case. There is more code, but the tests are more readable and assert more i.e how many times an expectation is called. IMHO, `.AnyTimes()` should be avoided when possible

Reply:
I refactored the test by adding a fixture to make it more readable, hopefully, so that the intent of each test case is not lost among `EXPECT` clutter

Reply:
Also, the tests check that `GetActiveSet` is called just once in each case except refetch (failed fetch / cancelation and then another handler call)

Reply:
> the intent of each test case is not lost among EXPECT clutter

The EXPECTs show the intent of the tests (what is _expected_ to happen) ;)

Reply:
Yes, I mean **repeated** large `EXPECT` constructs sometimes obscure the intent b/c of "just too much text". That needs some balance :)